aliasmatching/utils/utils.py

```diff
@@ -0,0 +1,265 @@
+
+        email_name_score = max(name_handle_dist((u1['first_name'], u1['last_name']),
+                                                u2['email']),
+                               name_handle_dist((u2['first_name'], u2['last_name']),
+                                                u1['email'])
+                               )
+
+        email_name_score = adjust_score(email_name_score, email_name_coef)
+
+        name_score = min(name_score, email_name_score)
+
+    # handle score
+
+    email_score = 1
+    if not u1['short_email'] is np.nan and not u2['short_email'] is np.nan:
+        if len(u1['short_email']) > 2 and len(u2['short_email']) > 2:
+            email_score = get_norm_levdist(u1['short_email'], u2['short_email'])
+    email_score = adjust_score(email_score, email_coef)
+
+    login_score = 1
+    if not u1['login'] is np.nan and not u2['login'] is np.nan:
+        if len(u1['login']) > 2 and len(u2['login']) > 2:
+            login_score = get_norm_levdist(u1['login'], u2['login'])
+    login_score = adjust_score(login_score, login_coef)
+
+    login_email_score = 1
+    if not u1['login'] is np.nan and not u2['short_email'] is np.nan:
+        if len(u1['login']) > 2 and len(u2['short_email']) > 2:
+            login_email_score = get_norm_levdist(u1['login'], u2['short_email'])
+
+    if not u1['short_email'] is np.nan and not u2['login'] is np.nan:
+        if len(u1['short_email']) > 2 and len(u2['login']) > 2:
+            login_email_score = min(login_email_score, get_norm_levdist(u1['short_email'], u2['login']))
+    login_email_score = adjust_score(login_email_score, login_email_coef)
+
+    login_name_score = max(name_handle_dist((u1['first_name'], u1['last_name']),
+                                            u2['login']),
+                           name_handle_dist((u2['first_name'], u2['last_name']),
+                                            u1['login'])
+                           )
+    login_name_score = adjust_score(login_name_score, login_name_coef)
+
+    handle_score = min(email_score, login_score, login_email_score, login_name_score)
+
+    return min(name_score, handle_score)
+
+
+def get_sim_matrix(users,
+                   name_coef=1,
+                   email_name_coef=1,
+                   email_score_coef=1,
+                   login_score_coef=1,
+                   login_email_coef=0,
+                   login_name_coef=0):
+    """
+    calculates name similarity matrix between users
+    """
+    sim_matrix = np.zeros((len(users), len(users)))
+    for i1, row1 in tqdm(users.iterrows()):
+        def score(i2, row2):
+            if i1 < i2:
+                return sim_users(row1, row2, name_coef, email_name_coef, email_score_coef,
+                                 login_score_coef, login_email_coef, login_name_coef)
+            if i1 == i2:
+                return 0
+            return 0
+
+        sim_matrix[i1] = [score(*p) for p in users.iterrows()]
```

Comment:
In my opinion, defining a method within another method increase the chance of making small mistakes and should be avoided unless there is no other easy way. I guess the following code will does the same job:

```suggestion
        sim_matrix[i1] = [
            sim_users(row1, row2, name_coef, email_name_coef, email_score_coef,login_score_coef, login_email_coef, login_name_coef)
                          for (i2, row2) in users.iterrows() if i1 < i2
        ]
```

Reply:
updated